# Hands-On Assignment 5

In this assignment, you will practice everything that you have learned so far in an end-to-end setting.
You will be provided with a dataset that is **unique to you**, and your task is to perform
all the steps from previous assignments to clean, explore, visualize, and analyze your dataset.

**Written Portion**: Additionally, you will create a report that describes your process and provides insights about your dataset.
Each section that should appear in your report is noted with an orange star (like normal HO tasks).  The report should be  4-6 pages (12 pt font, 1.5 line spacing), and turned in on Canvas as a PDF.

The coding aspect for this assignment will be turned in the same was as all other HO's,
by submitting this file to the autograder.


For this assignment, feel free to make additional functions instead of implementing everything in the provided function.

The objective of this assignment is for you to apply and solidify the skills you have learned in previous assignments.

# Prompt

You have graduated from this class, and are a huge success!
You landed a job doing data science at some fancy company.

You just got a new client with some really interesting problems you get to solve.
Unfortunately, because of a big mess-up on their side the data's metadata got corrupted
(and the person that used to maintain the data just took a vow of silence and moved to a bog).

The only column you are sure about is the `label` column,
which contains a numeric label for each row.
Aside from that, the client does not know anything about the names, content, or even data types for each column.

Your task is to explore, clean, and analyze this data.
You should have already received an email with the details on obtaining your unique data.
Place it in the same directory as this notebook (and your `local_grader.py` script) and name it `data.txt`.

*I know this prompt may sound unrealistic, but I have literally been in a situation exactly like this.
I was working at a database startup, and one of our clients gave us data with over 70 columns and more than a million records and told us:
"The person who used to manage the data is no longer working with us, but this was the data they used to make all their decisions.
We also lost all the metadata information, like column names."
...
Working in industry is not always glamorous.
-Eriq*

# Part 0: Explore Your Data

Before you start doing things to/with your data, it's always a good idea to load up your data and take a look.

In [44]:
import pandas
import numpy
import re
import matplotlib.pyplot as pyplot
import math
import sklearn
import scipy

# import sklearn.linear_model
# import sklearn.neighbors
# import sklearn.tree


# Modify this to point to your data.
unique_data = pandas.read_csv('data.txt', sep = "\t")
unique_data

label   col_00   col_01                                 col_02  col_03  \
0        0   850 kg   0.8561                    applied mathematics  Rabbit   
1        6  1028 kg   0.9142               Games and Playable Media     cat   
2        1  2401 kg    0.373                         Bioengineering     dog   
3        0   865 kg   1.8129                    Computational Media  rabbit   
4        5  1038 kg   0.6027            natural language processing    bird   
..     ...      ...      ...                                    ...     ...   
937      6   614 kg  -0.5986                 Electrical Engineering     cat   
938      6   488 kg   0.5701                       Computer Science     cat   
939      2  1849 kg   0.8054                   Robotics Engineering     pig   
940      3   496 kg   0.6137                          Biotechnology  ferret   
941      5  1716 kg   0.1899  Technology and Information Management    BIRD   

      col_04  col_05      col_06  col_07 col_08  col_09   col_10 col_11  \
0     0.4152  0.0210     HAMSTER  0.5747   2026  0.8048   1.6842    670   
1     0.6378     NaN     chicken  1.1363   -104 -1.2107  -1.4931    -91   
2      0.361  0.9553         cat -0.5059    705  3.0146   1.0999   1550   
3     0.9361 -0.0752     hamster  0.5208    410  1.6193  -0.6549    698   
4    -0.9037 -0.3321         pig -0.3035   -691  0.9618   -0.861   1747   
..       ...     ...         ...     ...    ...     ...      ...    ...   
937   0.5285  0.1782  guinea pig  1.3322    897  1.6021   0.6096    763   
938   0.0154 -0.0283         rat -0.0779   -266  0.8431  -0.3589     29   
939  -0.4352  1.2306      ferret -1.0848   1308 -0.0236   0.9168   -261   
940    0.258  0.7122         DOG  0.3552   1416 -0.0530   0.7684    771   
941  -0.1527  2.0224         pig  0.7081   -494  1.4933   0.3044    751   

      col_12 col_13  
0        rat    238  
1     rabbit   1018  
2       bird    303  
3        rat  -1114  
4    hamster    457  
..       ...    ...  
937   rabbit    526  
938      rat     14  
939     fish   1102  
940   ferret   -143  
941  hamster   1008  

[942 rows x 15 columns]

Don't forget to checkout the column information.

In [45]:
unique_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 942 entries, 0 to 941
Data columns (total 15 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   label   942 non-null    int64  
 1   col_00  936 non-null    object 
 2   col_01  934 non-null    object 
 3   col_02  940 non-null    object 
 4   col_03  935 non-null    object 
 5   col_04  933 non-null    object 
 6   col_05  930 non-null    float64
 7   col_06  931 non-null    object 
 8   col_07  939 non-null    float64
 9   col_08  938 non-null    object 
 10  col_09  936 non-null    float64
 11  col_10  933 non-null    object 
 12  col_11  935 non-null    object 
 13  col_12  930 non-null    object 
 14  col_13  938 non-null    object 
dtypes: float64(3), int64(1), object(11)
memory usage: 110.5+ KB


And any numeric information.

In [46]:
unique_data.describe()

label      col_05      col_07      col_09
count  942.000000  930.000000  939.000000  936.000000
mean     2.969214    0.642005    0.176143    0.438197
std      2.006924    0.854722    0.810569    0.972678
min      0.000000   -1.793300   -3.310000   -2.675400
25%      1.000000    0.052500   -0.379400   -0.218650
50%      3.000000    0.618600    0.165600    0.428350
75%      5.000000    1.227150    0.705800    1.121725
max      6.000000    3.043300    2.599400    3.174300

<h4 style="color: darkorange; font-size: x-large";>★ Written Task: Introduction</h4>

Briefly describe the dataset you’re given and define the goal of the project and how you approach it.
For example, you can present a basic introduction of your data (shape and proposed data types)
and your goal is to use these features to predict the label of the response variable.
Then you propose a few models that are suitable for this project which will be introduced in the modeling section.

# Part 1: Data Cleaning

As always, we should start with data cleaning.
Take what you learned from HO3 to clean up this messy data to a point where it is ready for machine learning algorithms.

Some things you may want to do:
 - Deal with missing/empty values.
 - Fix numeric columns so that they actually contain numbers.
 - Remove inconsistencies from columns.
 - Assign a data type to each column.

<h4 style="color: darkorange; font-size: x-large";>★ Task 1.A</h4>

Complete the following function that takes in a DataFrame and outputs a clean version of the DataFrame.
You can assume that the frame has all the same structure as your unique dataset.
You can return the same or a new data frame.

In [47]:
# Solely for surveying to determine best way to handle empty values
def print_empty_counts(frame):
    # print column counts
    for column in frame.columns:
        empty_count = frame[column].isna().sum()
        print(column, empty_count)

    # print row counts
    a = [0] * len(frame.columns)
    for index, row in frame.iterrows():
        empty_count = frame.loc[index].isna().sum()
        if empty_count > 0:
            a[empty_count] += 1
    print(a)


# Replaces entries that are the string "?" with numpy NaNs
def replace_question_marks_with_nans(frame):
    for column in frame.columns:
        for row_idx in frame.index:
            if frame.loc[row_idx, column] == '?':
                frame.loc[row_idx, column] = numpy.nan
    
    return frame

# Extracts numbers from strings and casts integer floats to ints
def extract_numbers(frame, ignore_columns = []):
    for column in frame.columns:
        if column in ignore_columns:
            continue
        if frame.dtypes[column] != 'object':  # skip uniform type columns
            continue

        # extract numbers from every row in column
        for i in frame.index:
            entry = frame.loc[i, column]
            if pandas.isna(entry):
                # value missing
                continue
            
            # replace strings with any number inside
            if isinstance(entry, str):
                # match numbers
                match = re.match(r"-?\d+(\.\d+)?", entry)
                if not match:
                    print("Failed to match: ", column, i, entry, match)

                # cast match to int or float
                number_str = match[0]
                if match[1]:  # if decimal found
                    entry = numpy.float64(number_str)
                else:
                    entry = numpy.int64(number_str)
                
                frame.loc[i, column] = entry
    
    return frame

# Attempts to cast column to its dominant (majority) type.
# If the dominant type is str, numbers are discarded
# If the type is numerical, strings are discarded and numbers casted.
def guess_types(frame):
    for column in frame.columns:
        # Count types in column
        str_count = frame[column].apply(lambda x: isinstance(x, str)).sum()
        int_count = frame[column].apply(lambda x: isinstance(x, (int, numpy.integer))).sum()
        # needed to shorten line length
        is_float = lambda x: isinstance(x, (float, numpy.floating))
        is_non_integer = lambda x: not x.is_integer() if is_float(x) else False
        float_count = frame[column].apply(lambda x: is_float(x) and is_non_integer(x)).sum()
        
        # Set col_type to dominant type
        col_type = None
        total_count = str_count + float_count + int_count
        threshold = .9 * total_count
        if str_count > threshold:
            col_type = str
        elif float_count > threshold:
            col_type = numpy.float64
        elif int_count > threshold:
            col_type = numpy.int64

        for row in frame.index:
            entry = frame.loc[row, column]
            if pandas.isna(entry):
                continue

            # Remove inconsistent or uncastable data types
            new_entry = entry
            if col_type is str:
                if not isinstance(entry, str):
                    new_entry = numpy.nan
            else:  # float64 or int64
                if isinstance(entry, str):
                    new_entry = numpy.nan
            
            frame.loc[row, column] = new_entry

        # Set column type
        if col_type is str:
            frame[column] = frame[column].astype("string")
        if col_type is numpy.float64:
            frame[column] = frame[column].astype("Float64")
        if col_type is numpy.int64:
            frame[column] = frame[column].astype("Int64")
        
    return frame


# Finds and returns outliers
def find_outliers(frame, deviations, label_column_name):
    outliers = dict()
    
    for column in frame.columns:
        if frame[column].dtype == "string" or frame[column].dtype == "object":
            continue
        
        sumx = 0
        sumx2 = 0
        n = 0
        hasStr = False
        for row in frame.index:
            entry = frame[column][row]
            if pandas.isna(entry):
                continue
            sumx += entry
            sumx2 += entry * entry
            n += 1

        if hasStr:
            continue
        
        outlier_list = []
        
        mean = sumx / n
        variance = sumx2 / n - mean ** 2
        stdev = math.sqrt(variance)
        for row in frame.index:
            entry = frame[column][row]
            if isinstance(entry, int) or isinstance(entry, float):
                if abs(entry - mean) > deviations * stdev:
                    outlier_list.append((frame[label_column_name][row], entry))

        if len(outlier_list) > 0:
            outliers[column] = outlier_list
    
    return outliers


def fill_missing_values(frame, ignore_columns = []):
    for column in frame.columns:
        if column in ignore_columns:
            continue

        column_type = frame.dtypes[column]
        column_mean = frame[column].mean()
        if column_type == "Int64":
            column_mean = round(column_mean)
        frame[column] = frame[column].fillna(column_mean)
        
    return frame


# Trims strings and makes them lowercase
def normalize_strings(frame):
    for column in frame.columns:
        if frame.dtypes[column] != "string":
            print(column)
            continue
        
        frame[column] = frame[column].apply(lambda x: x.strip().lower() if pandas.notna(x) else x)
    
    return frame


# One-hot encodes string columns
def one_hot(frame, column_name, suffix = ""):
    onehotcols = dict()
    i = 0
    for row in frame.index:
        entry = frame.at[row, column_name]
        if pandas.isna(entry):
            continue
        
        categories = entry.split(",")
        for category in categories:
            category = category.strip()
            
            # make sure category isnt empty
            if category == "":
                continue
            
            if category not in onehotcols:
                onehotcols[category + suffix] = [0] * len(frame)
            
            onehotcols[category + suffix][i] = 1

        i += 1
        
    frame.drop(column_name, axis=1, inplace=True)
    frame = pandas.concat([frame, pandas.DataFrame(onehotcols)], axis=1)

    return frame


def clean_data(frame):
    string_columns = ["col_02", "col_03", "col_06", "col_12"]

    print(1)
    frame = replace_question_marks_with_nans(frame)
    print(2)
    frame = extract_numbers(frame, ignore_columns = string_columns)
    print(3)
    frame = guess_types(frame)
    print(4)
    print("Outliers: ", find_outliers(frame, 4.0, 'label'))
    print(5)
    frame = fill_missing_values(frame, string_columns)
    print(6)
    frame = normalize_strings(frame)
    print(7)
    # print(frame["col_02"].unique())
    # print(frame["col_03"].unique())
    # print(frame["col_06"].unique())
    # print(frame["col_12"].unique())
    
    frame = one_hot(frame, "col_02")  # Major
    frame = one_hot(frame, "col_03", suffix = "_1")  # Pets 1
    frame = one_hot(frame, "col_06", suffix = "_2")  # Pets 2
    frame = one_hot(frame, "col_12", suffix = "_3")  # Pets 3
    
    return frame

unique_data = clean_data(unique_data)
unique_data

1
2
3
4
Outliers:  {'col_07': [(np.int64(2), np.float64(-3.31))]}
5
6
label
col_00
col_01
col_04
col_05
col_07
col_08
col_09
col_10
col_11
col_13
7


label  col_00  col_01  col_04    col_05  col_07  col_08  col_09  col_10  \
0        0     850  0.8561  0.4152     0.021  0.5747    2026  0.8048  1.6842   
1        6    1028  0.9142  0.6378  0.642005  1.1363    -104 -1.2107 -1.4931   
2        1    2401   0.373   0.361    0.9553 -0.5059     705  3.0146  1.0999   
3        0     865  1.8129  0.9361   -0.0752  0.5208     410  1.6193 -0.6549   
4        5    1038  0.6027 -0.9037   -0.3321 -0.3035    -691  0.9618  -0.861   
..     ...     ...     ...     ...       ...     ...     ...     ...     ...   
937      6     614 -0.5986  0.5285    0.1782  1.3322     897  1.6021  0.6096   
938      6     488  0.5701  0.0154   -0.0283 -0.0779    -266  0.8431 -0.3589   
939      2    1849  0.8054 -0.4352    1.2306 -1.0848    1308 -0.0236  0.9168   
940      3     496  0.6137   0.258    0.7122  0.3552    1416  -0.053  0.7684   
941      5    1716  0.1899 -0.1527    2.0224  0.7081    -494  1.4933  0.3044   

     col_11  ...  bird_3  hamster_3  reptile_3  pig_3  guinea pig_3  dog_3  \
0       670  ...       0          0          0      0             0      0   
1       -91  ...       0          0          0      0             0      0   
2      1550  ...       0          0          0      0             0      0   
3       698  ...       0          0          0      0             0      0   
4      1747  ...       0          0          0      0             0      0   
..      ...  ...     ...        ...        ...    ...           ...    ...   
937     763  ...       0          0          0      0             0      0   
938      29  ...       0          0          0      0             0      0   
939    -261  ...       0          0          0      0             0      0   
940     771  ...       0          0          0      0             0      0   
941     751  ...       0          0          0      0             0      0   

     cat_3  fish_3  chicken_3  ferret_3  
0        0       0          0         0  
1        0       0          0         0  
2        0       0          0         0  
3        0       0          0         0  
4        0       0          0         0  
..     ...     ...        ...       ...  
937      0       0          0         0  
938      0       0          0         0  
939      0       0          0         0  
940      0       0          0         0  
941      0       0          0         0  

[942 rows x 63 columns]

Now we should also be able to view all the numeric columns.

In [48]:
unique_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 942 entries, 0 to 941
Data columns (total 63 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   label                                  942 non-null    Int64  
 1   col_00                                 942 non-null    Int64  
 2   col_01                                 942 non-null    Float64
 3   col_04                                 942 non-null    Float64
 4   col_05                                 942 non-null    Float64
 5   col_07                                 942 non-null    Float64
 6   col_08                                 942 non-null    Int64  
 7   col_09                                 942 non-null    Float64
 8   col_10                                 942 non-null    Float64
 9   col_11                                 942 non-null    Int64  
 10  col_13                                 942 non-null    Int64  
 11  applie

<h4 style="color: darkorange; font-size: x-large";>★ Written Task: Data Cleaning</h4>

Describe the steps you took for data cleaning.
Why did you do this?
Did you have to make some choices along the way? If so, describe them.

# Part 2: Data Visualization

Once you have cleaned up the data, it is time to explore it and find interesting things.
Part of this exploration, will be visualizing the data in a way that makes it easier for yourself and others to understand.
Use what you have learned in HO1 and HO2 to create some visualizations for your dataset.

<h4 style="color: darkorange; font-size: x-large";>★ Written Task: Data Visualization</h4>

Create at least two different visualizations that help describe what you see in your dataset.
Include these visualizations in your report along with descriptions of
how you created the visualization,
what data preparation you had to do for the visualization (aside from the data cleaning in the previous part),
and what the visualization tells us about the data.

In [43]:
def plot_counts(frame, column):
    counts = frame[column].value_counts()
    pyplot.bar(counts.index, counts.values)
    pyplot.xticks(rotation=90)
    pyplot.title(f"Frequency of categories in {column}")
    pyplot.xlabel("Category")
    pyplot.ylabel("Frequency")
    pyplot.show()

def plot_hist(frame, column):
    pyplot.hist(frame[column], bins=10, edgecolor='black')
    pyplot.xlabel(column)
    pyplot.ylabel("Frequency")
    pyplot.title(f"Histogram of {column}")
    pyplot.show()

# for column in unique_data.columns:
#     print(column, unique_data.dtypes[column])
#     if unique_data.dtypes[column] == "string":
#         # plot_bar(unique_data, column)
#         continue

plot_counts(unique_data, "label")
# plot_counts(unique_data, "col_02")  # Majors
# plot_counts(unique_data, "col_03")  # Pets 1
# plot_counts(unique_data, "col_06")  # Pets 2
# plot_counts(unique_data, "col_12")  # Pets 3

label_filtered = unique_data[unique_data["label"] == 0]
# plot_counts(label_filtered, "col_02")  # Majors
# plot_counts(label_filtered, "col_03")  # Pets 1
# plot_counts(label_filtered, "col_06")  # Pets 2
# plot_counts(label_filtered, "col_12")  # Pets 3

# for column in label_filtered.columns:
#     coltype = label_filtered.dtypes[column]
#     print(coltype)
#     if coltype != "Int64" and coltype != "Float64":
#         continue
#     plot_hist(label_filtered, column)

# for i in range(0, 7): #[0, 1]:  # see: col_04, [0, 1, 3, 5]
#     plot_hist(unique_data[unique_data["label"] == i], "col_04")
# for i in [3, 5]:  # see: col_04, [0, 1, 3, 5]
#     plot_hist(unique_data[unique_data["label"] == i], "col_04")

# plot_xy(frame,  4)

KeyError: 'col_02'

# Part 3: Modeling

Now that you have a good grasp of your clean data,
it is time to do some machine learning!
(Technically all our previous steps were also machine learning,
but now we get to use classifiers!)

Use the skills you developed to select **three** classifiers and implement them on your data.
For example, you can narrow down your choices to three classifiers which may include:
- Logistic regression
- K-nearest neighbors
- Decision tree
- Or others

<h4 style="color: darkorange; font-size: x-large";>★ Task 3.A</h4>

Complete the following function that takes in no parameters,
and returns a list with **three** untrained classifiers you are going to explore in this assignment.
This method may set parameters/options for the classifiers, but should not do any training/fitting.

For example, if you wanted to use logistic regression,
then **one** of your list items may be:
```
sklearn.linear_model.LogisticRegression()
```

In [ ]:
def create_classifiers():
    return [
        sklearn.linear_model.LogisticRegression(),
        sklearn.neighbors.KNeighborsClassifier(n_neighbors = 10),
        sklearn.tree.DecisionTreeClassifier()
    ]

my_classifiers = create_classifiers()
my_classifiers

Now that we have some classifiers, we can see how they perform.

<h4 style="color: darkorange; font-size: x-large";>★ Task 3.B</h4>

Complete the following function that takes in an untrained classifier, a DataFrame, and a number of folds.
This function should run k-fold cross validation with the classifier and the data,
and return a list with the accuracy of each run of cross validation.
You can assume that the frame has the column `label` and the rest of the columns can be considered clean numeric features.

Note that you may have to break your frame into features and labels to do this.
Do not change the passed-in frame (make copies instead).

If you are getting any `ConvergenceWarning`s you may either ignore them,
or try and address them
(they will not affect your autograder score, but may be something to discuss in the written portion of this assignment).

In [ ]:
def cross_fold_validation(classifier, frame, folds):
    y = frame["label"]
    x = frame.drop(columns = "label")  # creates a copy

    # scikit-learn functions
    # on KFold: use random_state parameter for reproducibility
    k_fold = sklearn.model_selection.KFold(n_splits = folds, shuffle = True)
    accuracy_scores = sklearn.model_selection.cross_val_score(classifier, x, y, cv = k_fold)
    
    return list(accuracy_scores)


my_classifiers_scores = []
for classifier in my_classifiers:
    accuracy_scores = cross_fold_validation(classifier, unique_data, 5)
    my_classifiers_scores.append(accuracy_scores)
    print("Classifier: %s, Accuracy: %s." % (type(classifier).__name__, accuracy_scores))

    print("Mean Accuracy: ", sum(accuracy_scores) / len(accuracy_scores))

<h4 style="color: darkorange; font-size: x-large";>★ Task 3.C</h4>

Complete the following function that takes in two equally-sized lists of numbers and a p-value.
This function should compute whether there is a statistical significance between
these two lists of numbers using a [Student's t-test](https://en.wikipedia.org/wiki/Student%27s_t-test)
at the given p-value.
Return `True` if there is a statistical significance, and `False` otherwise.
Hint: If you wish, you may use the `ttest_ind()` [method](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html) provided in the scipy package. 

In [37]:
def significance_test(a_values, b_values, p_value):
    # did not adjust variance, so set equal_var to false
    t_statistic, p = scipy.stats.ttest_ind(a_values, b_values, equal_var = False)
    return p < p_value

for i in range(len(my_classifiers)):
    for j in range(i + 1, len(my_classifiers)):
        significant = significance_test(my_classifiers_scores[i], my_classifiers_scores[j], 0.10)
        print("%s vs %s: %s" % (type(my_classifiers[i]).__name__,
                                type(my_classifiers[j]).__name__, significant))

LogisticRegression vs KNeighborsClassifier: True
LogisticRegression vs DecisionTreeClassifier: True
KNeighborsClassifier vs DecisionTreeClassifier: True


<h4 style="color: darkorange; font-size: x-large";>★ Written Task: Modeling</h4>

Describe the classifiers you have chosen.
Be sure to include all details about any parameter settings used for the algorithms.

Compare the performance of your models using k-fold validation.
You may look at accuracy, F1 or other measures.

Then, briefly summarize your results.
Are your results statistically significant?
Is there a clear winner?
What do the standard deviations look like, and what do they tell us about the different models?
Include a table like Table 1.

<center>Table 1: Every table needs a caption.</center>

| Model | Mean Accuracy | Standard Deviation of Accuracy |
|-------|---------------|--------------------------------|
| Logistic Regression | 0.724 | 0.004
| K-Nearest Neighbor | 0.750 | 0.003
| Decision Tree | 0.655 | 0.011

# Part 4: Analysis

Now, take some time to go over your results for each classifier and try to make sense of them.
 - Why do some classifiers work better than others?
 - Would another evaluation metric work better than vanilla accuracy?
 - Is there still a problem in the data that should fixed in data cleaning?
 - Does the statistical significance between the different classifiers make sense?
 - Are there parameters for the classifier that I can tweak to get better performance?

<h4 style="color: darkorange; font-size: x-large";>★ Written Task: Analysis</h4>

Discuss your observations, the relationship you found, and how you applied concepts from the class to this project.
For example, you may find that some feature has the most impact in predicting your response variable or removing a feature improves the model accuracy.
Or you may observe that your training accuracy is much higher than your test accuracy and you may want to explain what issues may arise.

# Part 5: Conclusion

<h4 style="color: darkorange; font-size: x-large";>★ Written Task: Conclusion</h4>

Briefly summarize the important results and conclusions presented in the project.
What are the important points illustrated by your work?
Are there any areas for further investigation or improvement?

<h4 style="color: darkorange; font-size: x-large";>★ Written Task: References</h4>

Include a standard bibliography with citations referring to techniques or published papers you used throughout your report (if you used any).

For example:
```
[1] Derpanopoulos, G. (n.d.). Bayesian Model Checking & Comparison.
https://georgederpa.github.io/teaching/modelChecking.html.
```

# Part XC: Extra Credit

So far you have used a synthetic dataset that was created just for you.
But, data science is always more interesting when you are dealing with actual data from the real world.
Therefore, you will have an opportunity for extra credit on this assignment using real-world data.

For extra credit, repeat the **written tasks** of Parts 0 through 4 with an additional dataset that you find yourself.
For the written portion of the extra credit for Part 0, include information about where you got the data and what the data represents.
You may choose any dataset that represents real data (i.e., is **not** synthetic or generated)
and is **not** [pre-packaged in scikit-learn](https://scikit-learn.org/stable/datasets.html).

Below are some of the many places you can start looking for datasets:
 - [Kaggle](https://www.kaggle.com/datasets) -- Kaggle is a website focused around machine learning competitions,
       where people compete to see who can get the best results on a dataset.
       It is very popular in the machine learning community and has thousands of datasets with descriptions.
       Make sure to read the dataset's description, as Kaggle also has synthetic datasets.
 - [data.gov](https://data.gov/) -- A portal for data from the US government.
        The US government has a lot of data, and much of it has to be available to the public by law.
        This portal contains some of the more organized data from several different government agencies.
        In general, the government has A LOT of interesting data.
        It may not always be clean (remember the CIA factbook), but it is interesting and available.
        All data here should be real-world, but make sure to read the description to verify.
 - [UCI's Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets.php) -- UC Irvine has their own data repository with a few hundred datasets on many different topics.
        Make sure to read the dataset's description, as UCI also has synthetic datasets.
 - [WHO's Global Health Observatory](https://apps.who.int/gho/data/node.home) -- The World Health Organization keeps track of many different health-related statistics for most of the countries in the world.
        All data here should be real-world, but make sure to read the description to verify.
 - [Google's Dataset Search](https://datasetsearch.research.google.com/) -- Google indexes many datasets that can be searched here.

You can even create a dataset from scratch if you find some data you like that is not already organized into a specific dataset.
The only real distinction between "data" and a "dataset" is that a dataset is organized and finite (has a fixed size).

Create a new section in your written report for this extra credit and include all the written tasks for the extra credit there.
Each written task/section that you complete for your new dataset is eligible for extra credit (so you can still receive some extra credit even if you do not complete all parts).
There is no need to submit any code for the extra credit.
If you created a new dataset, include the dataset or links to it with your submission.